project/src/com/google/daggerqueryui/scripts/binding_graph.js

```diff
@@ -238,6 +250,31 @@ const bindingGraph = (function() {
       }
     },
 
+    /**
+     * Adds all nodes which depend on the source node.
+     *
+     * @param {string} source
+     * @param {string[]} ancestors
+     */
+    addAncestors: function (source, ancestors) {
+      for (const ancestorNode of ancestors) {
```

Comment:
Because of the `O(n)` `hasEdge` check within `addEdge`, the `addAncestors` operation now becomes `O(n^2)`. This may be fine for the demo though since the graph is small. If you run into performance issues later, you might want to optimise the `hasEdge` function, say keeping a copy of the edges in a separate `Set` (though this means 2x used memory as well).

Reply:
Yes, I agree with you, it's a problem. I am also not happy with calls like `getAllEdges`, `getAllEdgesToDeps` when they are used only for understanding it there are such edge or not.

In the next PR I will add adjacency list and save everything in it (it's better then set of edges since we can easily understand is there an edge which ends in the specified node, for example). I [have created a card ](https://github.com/googleinterns/dagger-query/projects/2#card-45756862)for this issue.